modules/alookup/alookup_test.go

```diff
@@ -257,9 +349,140 @@ func TestDoLookup(t *testing.T) {
 		Flags:       miekg.DNSFlags{},
 	}
 
-	res, _, status, _ = l.DoLookup("example.com", "")
-	if status != zdns.STATUS_NO_ANSWER {
-		t.Errorf("Expected NO_ANSWER status, got %v", status)
+	res, _, status, _ := l.DoLookup("example.com", "")
+	if status != zdns.STATUS_NOERROR {
+		t.Errorf("Expected NOERROR status, got %v", status)
+	} else if res != nil {
+		t.Errorf("Expected no results, got %v", res)
+	}
+}
+
+func TestMismatchIpType(t *testing.T) {
+	_, glf, _, l := InitTest()
+	glf.IPv4Lookup = true
+	glf.IPv6Lookup = true
+	mockResults["example.com"] = miekg.Result{
+		Answers: []interface{}{miekg.Answer{
+			Ttl:    3600,
+			Type:   "A",
+			Class:  "IN",
+			Name:   "example.com",
+			Answer: "2001:db8::4",
+		}},
+		Additional:  nil,
+		Authorities: nil,
+		Protocol:    "",
+		Flags:       miekg.DNSFlags{},
+	}
+
+	res, _, status, _ := l.DoLookup("example.com", "")
+	if status != zdns.STATUS_ERROR {
+		t.Errorf("Expected NOERROR status, got %v", status)
+	} else if res != nil {
+		t.Errorf("Expected no results, got %v", res)
+	}
+}
+
+func TestCnameLoops(t *testing.T) {
+	_, glf, _, l := InitTest()
+	glf.IPv4Lookup = true
+	mockResults["cname.example.com"] = miekg.Result{
+		Answers: []interface{}{miekg.Answer{
+			Ttl:    3600,
+			Type:   "CNAME",
+			Class:  "IN",
+			Name:   "cname.example.com",
+			Answer: "example.com.",
+		}},
+		Additional:  nil,
+		Authorities: nil,
+		Protocol:    "",
+		Flags:       miekg.DNSFlags{},
+	}
+	mockResults["example.com"] = miekg.Result{
+		Answers: []interface{}{miekg.Answer{
+			Ttl:    3600,
+			Type:   "CNAME",
+			Class:  "IN",
+			Name:   "example.com",
+			Answer: "cname.example.com.",
+		}},
+		Additional:  nil,
+		Authorities: nil,
+		Protocol:    "",
+		Flags:       miekg.DNSFlags{},
+	}
+
+	res, _, status, _ := l.DoLookup("example.com", "")
+	if status != zdns.STATUS_ERROR {
+		t.Errorf("Expected ERROR status, got %v", status)
+	} else if res != nil {
+		t.Errorf("Expected no results, got %v", res)
+	}
+}
+
+func TestEmptyNonTerminal(t *testing.T) {
+	_, glf, _, l := InitTest()
+	glf.IPv4Lookup = true
+	mockResults["leaf.intermediate.example.com"] = miekg.Result{
+		Answers: []interface{}{miekg.Answer{
+			Ttl:    3600,
+			Type:   "A",
+			Class:  "IN",
+			Name:   "leaf.intermediate.example.com",
+			Answer: "192.0.2.3",
+		}},
+		Additional:  nil,
+		Authorities: nil,
+		Protocol:    "",
+		Flags:       miekg.DNSFlags{},
+	}
+	mockResults["intermediate.example.com"] = miekg.Result{
+		Answers:     nil,
+		Additional:  nil,
+		Authorities: nil,
+		Protocol:    "",
+		Flags:       miekg.DNSFlags{},
+	}
+	// Verify leaf returns correctly
+	res, _, _, _ := l.DoLookup("leaf.intermediate.example.com", "")
+	verifyResult(t, res.(Result), []string{"192.0.2.3"}, nil)
+
+	// Verify empty non-terminal returns no answer
+	res, _, status, _ := l.DoLookup("intermediate.example.com", "")
+	if status != zdns.STATUS_NOERROR {
```

Comment:
You also need to check that no answers were returned. The NOERROR does not mean no answers were returned.

Reply:
That is done in the next line.